### PR TITLE
RMC-134 Large Print Questionnaires and Translation Booklet Fulfilment Requests

### PR DESCRIPTION
**Pull request by AdamHawtin**
# Motivation and Context
Action scheduler needs to handle fulfilment requests for large print questionnaires and translation booklets which should _not_ request new UAC/QID pairs.

# What has changed
* Determine action type from fulfilment code
* Don't request UAC QID pair for TB/LP fulfilment action types
* Add unit tests for all P_OR_HX action types
* Add unit and integration tests for examples of new TB and LP fulfilment codes
* Make case ID a UUID in the fulfilment request DTO (was already required but will now just fail earlier if invalid and needs less string conversions)

# How to test?
Build and run this branch along with the print file service branch the run the linked acceptance test branch.

# Links
https://trello.com/c/yDd9LhBR/1009-rmc-134c2-implement-large-print-and-translations-requested-event-5
https://github.com/ONSdigital/census-rm-acceptance-tests/pull/108
https://github.com/ONSdigital/census-rm-print-file-service/pull/10

### CODE REVIEW

**Comment by AdamHawtin** [comment]
[RMC-134(C2) Implement Large Print and Translations requested event (5)](https://trello.com/c/yDd9LhBR/1009-rmc-134c2-implement-large-print-and-translations-requested-event-5)

File: src/test/java/uk.gov.ons.census.action/messaging/FulfilmentRequestReceiverIT.java [hunk 1]
```diff
@@ -83,13 +85,79 @@ public void checkReceivedEventsAreEmitted() throws InterruptedException, IOExcep
...
+        actionFulfilmentEvent.getPayload().getFulfilmentRequest().getContact(),
+        actualPrintFileDto);
+    assertThat(actualPrintFileDto.getUac()).isNull();
+    assertThat(actualPrintFileDto.getQid()).isNull();
+  }
+
+  private void checkAddressFieldsMatch(
+      Case expectedCase, Contact expectedContact, PrintFileDto actualPrintFileDto) {
+    assertThat(actualPrintFileDto)
+        .isEqualToComparingOnlyGivenFields(
```
Comment (liamtoozer) [inline src/test/java/uk.gov.ons.census.action/messaging/FulfilmentRequestReceiverIT.java]: Nice!

File: src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+  @Transactional
+  @ServiceActivator(inputChannel = "actionFulfilmentInputChannel")
+  public void receiveEvent(ResponseManagementEvent event) {
+    Case fulfilmentCase = fetchFulfilmentCase(event);
+    String fulfilmentCode = event.getPayload().getFulfilmentRequest().getFulfilmentCode();
+
+    Optional<ActionType> actionType;
+    try {
+      actionType = determineActionType(fulfilmentCode);
+    } catch (UnknownFulfilmentException e) {
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java]: Noooooo. Never never never use checked exceptions. Completely unnecessary.

File: src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+      log.with("fulfilmentCode", fulfilmentCode).warn("Unexpected fulfilment code received");
+      return;
+    }
+    if (actionType.isEmpty()) {
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java]: whaaat? Looks like we're handing things in a million different ways? Why do the stuff above if we're just going to do this?

File: src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      case "UACHHT2":
+      case "UACHHT2W":
+      case "UACHHT4":
+      case "UACIT1":
+      case "UACIT2":
+      case "UACIT2W":
+      case "UACIT4":
+        return Optional.empty(); // Ignore SMS fulfilments
+      default:
+        throw new UnknownFulfilmentException(fulfilmentCode);
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java]: nope. If a runtime exception is what we want, throw one here. If we want to ignore it return empty. We're going down the road of mountains of YAGNI code.

File: src/main/java/uk/gov/ons/census/action/messaging/UnknownFulfilmentException.java [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.ons.census.action.messaging;
+
+import lombok.AllArgsConstructor;
+
+@AllArgsConstructor
+public class UnknownFulfilmentException extends Exception {
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/action/messaging/UnknownFulfilmentException.java]: Do not need. Never use checked exceptions. Not useful.

File: src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+    rabbitTemplate.convertAndSend(outboundExchange, outboundPrinterRoutingKey, printFileDto);
+  }
+
+  private Case fetchFulfilmentCase(ResponseManagementEvent event) {
+    UUID caseId = event.getPayload().getFulfilmentRequest().getCaseId();
+    Optional<Case> fulfilmentCase = caseRepository.findByCaseId(caseId);
+    log.with("caseId", caseId)
+        .with("fulfilmentCode", event.getPayload().getFulfilmentRequest().getFulfilmentCode())
+        .debug("Fulfilment Requested Event");
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java]: Not "Fulfilment Requested Event". More like "inside random private method".

File: src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java (hunk 4 above)
Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java]: if you want to log stuff, here would be a good place to do it.

File: src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+      case "UACIT4":
+        return Optional.empty(); // Ignore SMS fulfilments
+      default:
+        throw new UnknownFulfilmentException(fulfilmentCode);
+    }
+  }
+
+  private static final Map<String, Integer> fulfilmentCodeToQuestionnaireType =
+      Map.of(
+          "P_OR_H1", 1,
```
Comment (NickGrantONS) [inline src/main/java/uk/gov/ons/census/action/messaging/FulfilmentRequestReceiver.java]: I know you just renamed the class, but @LukeLoze1 hates magic numbers so we could replace these with some constants while we're messing around with this class.

**Pull request merged**